Issue #604 - Add support for nominal forecasts

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #604.

Nominal forecasts are forecasts for outcomes that can fall in one of several unordered categories. This PR implements support for nominal forecasts (see #604, #607, and #608). 

Specifically, the PR
- creates a new `nominal_forecast` class with
  - an `assert_input_nominal` function that checks the inputs passed to a scoring function
  - a `check_input_nominal`, doing the same thing without producing an error - UPDATE: I think I deleted that as I didn't use it for checks. See #840 for some discussion on when to check what.
  - an `assert_forecast.forecast_nominal` function, checking that a data.table is complying with the required input format
  - a default list of metrics, provided via `metrics_nominal`
  - a new method `score.forecast_nominal`
- adds new example data
- updates `as_forecast()` to accept a new `predicted_label` argument.
- updates `get_forecast_type()` and adds a check function to make sure that the forecast type is nominal
- implements the log score for nominal forecasts
- adds tests

Note: 
Throughout the process, I noticed that sadly, scoringutils is currently _not_ "easily extensible"... To make this go smoothly, there are quite a few hoops. Some of this will be simplified in the future when we implement a separate `as_forecast_nominal()` function instead of a single `as_forecast()` function that has to do all the guesswork. 


---
Still missing (likely for a future PR)
- Updating the manuscript to include nominal forecasts
- other kinds of docs
  - Creating a vignette that walks through a hubVerse example
- A helper function that completes the forecast such that users don't have to specify every single option (see #608)

One current code example: 

```
# remotes::install_github("epiforecasts/scoringutils@multiclass")
library(dplyr)
library(hubExamples)
library(scoringutils)

pred <- hubExamples::forecast_outputs |> filter(output_type == "pmf")
obs <- hubExamples::forecast_target_observations |> 
  dplyr::filter(output_type == "pmf")
hubex <- dplyr::full_join(pred, obs)

hubex |> 
  dplyr::group_by(model_id, location, reference_date, horizon, target_end_date, target, output_type) |>
  dplyr::mutate(
    observation = output_type_id[observation == 1], 
    observation = factor(observation, levels = c("low", "moderate", "high", "very high")), 
    output_type_id = factor(output_type_id, levels =  c("low", "moderate", "high", "very high"))) |>
  as_forecast(
    model = "model_id", observed = "observation", 
    predicted = "value", predicted_label = "output_type_id"
  ) |> 
  score()
```


## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

seabbs commented:
@nikosbosse do you want me to look at this? Looks stuck right now

nikosbosse commented:
Main reason it's stuck is because I haven't added enough tests. Let me fix things up first and then you can have a look, I think that makes more sense

nickreich commented on R/forecast.R:
```diff
@@ -178,6 +178,34 @@ as_forecast_sample <- function(data,
+
+#' @rdname as_forecast
+#' @param predicted_label (optional) Name of the column in `data` that denotes
+#'   the outcome to which a preidcted probability corresponds to.
```
"preidcted" --> "predicted"

nickreich commented on tests/testthat/test-forecast.R:
```diff
@@ -157,6 +157,35 @@ test_that("output of as_forecasts() is accepted as input to score()", {
 
+
+# as_forecast.forecast_nominal() -----------------------------------------------
+test_that("as_forecast.forecast_nominal() works as expected", {
```
I think this test is checking that all forecasts have all necessary rows, even if predicted probabilities is zero. maybe that should be a more specific description of the test here 
```
test_that("as_forecast.forecast_nominal() breaks when rows with zero probability are missing")
```

seabbs replied:
Note for the future that you can use the suggestion interface (the piece of paper in the comment bar with +- on it) to make this a code suggestion that can be batched merged.

nickreich reviewed (approved):
I didn't review closely a lot of the code related to formal S3 class setup because i'm not that familiar with the structure/functions used there. but I reviewed the tests and the general set-up with the nominal forecast type and things look good to me +/- a few very small optional suggested changes.

seabbs commented on R/check-inputs-scoring-functions.R:
```diff
@@ -193,6 +193,57 @@ check_input_binary <- function(observed, predicted) {
+#' @title Assert that inputs are correct for nominal forecasts
+#' @description Function assesses whether the inputs correspond to the
+#' requirements for scoring nominal forecasts.
+#' @param observed XXX
```
Docs are expected to be inherited here or otherwise missed?

seabbs replied:
This is the only point that I think needs to be addressed in this PR

nikosbosse replied:
Updated the documentation

seabbs commented on R/get_-functions.R:
```diff
@@ -16,7 +16,9 @@
 get_forecast_type <- function(data) {
   assert_data_frame(data)
   assert(check_columns_present(data, c("observed", "predicted")))
-  if (test_forecast_type_is_binary(data)) {
```
So this is a remaining pain point for the package right in terms of adding a new class. Do we have a specific issue for this? Seems tricky to resolve unless we can just rely on s3 dispatch?

We are using this as a check_forecast_type mostly right?

nikosbosse replied:
I think we can simplify `get_forecast_type()` or even get rid of it entirely. Now that we have `as_forecast_<type>` I don't think we ever need to check for a forecast type without knowing the class of the object we're checking.

nikosbosse replied:
Issue: https://github.com/epiforecasts/scoringutils/issues/887

seabbs commented on R/get_-functions.R:
```diff
@@ -262,7 +277,7 @@ get_protected_columns <- function(data = NULL) {
 
   protected_columns <- c(
```
this is another pain point. I guess we could have a s3 method for all classes that lists the protected columns for that ? Are we using this internally anywhere we don't know about the class of the input the user has made? If no we could do this and make this a lot more modular/hard coded

nikosbosse replied:
Issue: https://github.com/epiforecasts/scoringutils/issues/889

seabbs commented on R/get_-functions.R:
```diff
@@ -310,8 +325,8 @@ get_duplicate_forecasts <- function(
   assert_data_frame(data)
   data <- ensure_data.table(data)
   forecast_unit <- get_forecast_unit(data)
-  available_type <- c("sample_id", "quantile_level") %in% colnames(data)
```
Again if we can rely on the input being a forecast object we can make this s3 and avoid a new class requiring hard coded changes to the package

nikosbosse replied:
Issue: https://github.com/epiforecasts/scoringutils/issues/888

seabbs reviewed (approved):
This looks really good I think and also appears correct to me. I don't have substantive comments about this PR. 

I did however use it to review the current changes needed to add a new class. This has improved by splitting out as_forecast but there are still a few pain points. It looks like nearly all of the me can deal with using a bit more s3 which is great. 

I think we have discussed this before but I think this would be much easier to review/parse and easier for someone new to do if all the bits that defined as specific as_forecast_type where in the same file vs being split by generic method.

nikosbosse commented:
@seabbs some excellent points in your review here. I think moving towards `as_forecast_<type>()` really was the right call and should allow us to simplify things here quite a bit. 

I suggest addressing your points before implementing ordinal forecasts (pinging @nickreich and @elray1 for awareness) as that will make it easier to create the new ordinal class. Since Nick and Evan care about the ordinal forecasts more than the nominal ones I also suggest addressing your points before merging this.

seabbs commented:
>  I also suggest addressing your points before merging this.

I don't mind either way here but I agree it would be a good idea to use the ordinal forecasts as a test case. If it were me I think I would look to merge this, make a new issue with the pain points identified, address in a PR, and then implement nominal?
